src/ParameterRegistry.cpp

```diff
@@ -0,0 +1,600 @@
+
+#include "ParameterRegistry.h"
+#include "Logger.h"
+#include "userConfig.h"
+#include <algorithm>
+
+ParameterRegistry ParameterRegistry::_singleton;
+
+// Global variables, needed for backwards compatibility
+extern uint8_t pidON;
+extern uint8_t usePonM;
+extern double aggKp;
+extern double aggTn;
+extern double aggTv;
+extern double aggIMax;
+extern double steamKp;
+extern double brewSetpoint;
+extern double brewTempOffset;
+extern double brewPIDDelay;
+extern uint8_t useBDPID;
+extern double aggbKp;
+extern double aggbTn;
+extern double aggbTv;
+extern double emaFactor;
+extern double steamSetpoint;
+extern double targetBrewTime;
+extern double preinfusion;
+extern double preinfusionPause;
+extern double targetBrewWeight;
+extern int backflushCycles;
+extern double backflushFillTime;
+extern double backflushFlushTime;
+extern uint8_t standbyModeOn;
+extern double standbyModeTime;
+extern uint8_t featureBrewControl;
+extern uint8_t featureFullscreenBrewTimer;
+extern uint8_t featureFullscreenManualFlushTimer;
+extern double postBrewTimerDuration;
+extern uint8_t featureHeatingLogo;
+extern uint8_t featurePidOffLogo;
+extern float scaleCalibration;
+extern float scale2Calibration;
+extern float scaleKnownWeight;
+extern uint8_t steamON;
+extern uint8_t backflushOn;
+extern double temperature;
+extern uint8_t scaleTareOn;
+extern uint8_t scaleCalibrationOn;
+extern int logLevel;
+extern const char sysVersion[64];
+
+void ParameterRegistry::initialize(Config& config) {
+    if (_ready) {
+        return;
+    }
+
+    _config = &config;
+
+    _parameters.clear();
+    _parameterMap.clear();
+    _pendingChanges = false;
+    _lastChangeTime = 0;
+
+    // Add all parameters
+    auto addParam = [this](const std::shared_ptr<Parameter>& param) {
+        _parameters.push_back(param);
+        _parameterMap[param->getId()] = param;
+    };
+
+    // PID Section
+    addParam(std::make_shared<Parameter>(
+        "PID_ON", "Enable PID Controller", kUInt8, sPIDSection, 1,
+        [&config]() -> bool {
+            pidON = config.getPidEnabled();
+            return pidON;
+        },
+        [&config](const bool val) {
+            config.setPidEnabled(val);
+            pidON = val;
+        },
+        false, "", [] { return true; }, &pidON));
+
+    addParam(std::make_shared<Parameter>(
+        "PID_USE_PONM", "Enable PonM", kUInt8, sPIDSection, 2,
+        [&config]() -> bool {
+            usePonM = config.getUsePonM();
+            return usePonM;
+        },
+        [&config](const bool val) {
+            config.setUsePonM(val);
+            usePonM = val;
+        },
+        true, "Use PonM mode (<a href='http://brettbeauregard.com/blog/2017/06/introducing-proportional-on-measurement/' target='_blank'>details</a>)", [] { return true; }, &usePonM));
+
+    addParam(std::make_shared<Parameter>(
```

Comment:
I would suggest changing the `addParam` to accept variadic arguments and do the `make_shared` inside the function:

```cpp
template <typename... Args>
void addParam(Args&&... args) {
  _parameters.emplace_back(std::forward<Args>(args)...));
}
```